src/providers/mod.rs

```diff
@@ -0,0 +1,124 @@
+use failure::Error;
+use std::collections::HashMap;
+
+#[derive(Debug)]
+pub struct LaunchDescriptor<M: MachineSetup> {
+    pub region: M::Region,
+    pub log: slog::Logger,
+    pub max_wait: Option<std::time::Duration>,
+    pub machines: Vec<(String, M)>,
+}
+
+/// This is used to group machines into connections
+/// to cloud providers. e.g., for AWS we need a separate
+/// connection to each region.
+pub trait MachineSetup {
+    type Region: Eq + std::hash::Hash + Clone + ToString;
+    fn region(&self) -> Self::Region;
+}
+
+/// Implement this trait to implement a new cloud provider for Tsunami.
+/// Tsunami will call `launch` once per unique region, as defined by `MachineSetup`.
+pub trait Launcher {
+    type MachineDescriptor: MachineSetup;
+
+    /// Spawn the instances. Implementors should remember enough information to subsequently answer
```

Comment:
I think a bit more detail may be useful here. For example the fact that this method may be called multiple times.